ext/avromatic/decoder.cpp

```diff
@@ -0,0 +1,192 @@
+    }
+    case avro::AVRO_LONG:
+    {
+      if (datum.logicalType().type() == avro::LogicalType::TIMESTAMP_MILLIS) {
+        rb_value = deserialize_logical_timestamp_millis(datum.value<long>());
+      } else if (datum.logicalType().type() == avro::LogicalType::TIMESTAMP_MICROS) {
+        rb_value = deserialize_logical_timestamp_micros(datum.value<long>());
+      } else {
+        rb_value = LONG2NUM(datum.value<long>());
+      }
+      break;
+    }
+    case avro::AVRO_FLOAT:
+    {
+      rb_value = DBL2NUM(datum.value<float>());
+      break;
+    }
+    case avro::AVRO_DOUBLE:
+    {
+      rb_value = DBL2NUM(datum.value<double>());
+      break;
+    }
+    case avro::AVRO_BOOL:
+    {
+      rb_value = datum.value<bool>() ? Qtrue : Qfalse;
+      break;
+    }
+    case avro::AVRO_NULL:
+    {
+      rb_value = Qnil;
+      break;
+    }
+    case avro::AVRO_RECORD:
+    {
+      rb_value = rb_hash_new();
+      const avro::GenericRecord& record_value = datum.value<avro::GenericRecord>();
+      const int num_fields = (int)record_value.fieldCount();
+      for (int i=0; i<num_fields; ++i) {
+        const std::string& field_name = record_value.schema()->nameAt(i);
+        VALUE rb_field_name = ID2SYM(rb_intern(field_name.c_str()));
+        const avro::GenericDatum& field_datum = record_value.fieldAt(i);
+        rb_hash_aset(rb_value, rb_field_name, datum_to_ruby_value(record_value.schema()->leafAt(i), field_datum, strict));
+      }
+      break;
+    }
+    case avro::AVRO_ENUM:
+    {
+      const std::string& symbol = datum.value<avro::GenericEnum>().symbol();
+      rb_value = ID2SYM(rb_intern(symbol.c_str()));
+      break;
+    }
+    case avro::AVRO_ARRAY:
+    {
+      const avro::GenericArray::Value& array_value = datum.value<avro::GenericArray>().value();
+      const long num_elements = array_value.size();
+      rb_value = rb_ary_new_capa(num_elements);
+      for (long i=0; i<num_elements; ++i) {
+        rb_ary_store(rb_value, i, datum_to_ruby_value(schema->leafAt(0), array_value[i], strict));
+      }
+      break;
+    }
+    case avro::AVRO_MAP:
+    {
+      rb_value = rb_hash_new();
+      const avro::GenericMap::Value& map_entries = datum.value<avro::GenericMap>().value();
+      for (auto iter = map_entries.cbegin(); iter != map_entries.cend(); ++iter) {
+        VALUE rb_map_key = utf8_string_to_ruby(iter->first);
+        VALUE rb_map_value = datum_to_ruby_value(schema->leafAt(1), iter->second, strict);
+        rb_hash_aset(rb_value, rb_map_key, rb_map_value);
+      }
+     break;
+    }
+    case avro::AVRO_FIXED:
+    {
+      const avro::GenericFixed& fixedValue = datum.value<avro::GenericFixed>();
+      const std::vector<uint8_t>& bytes = fixedValue.value();
+      rb_value = rb_str_new((const char*)&bytes[0], bytes.size());
+      break;
+    }
+    default:
+      rb_raise(rb_eRuntimeError, "Unexpected field type %s", avro::toString(datum.type()).c_str());
+      // Make the compiler happy
```

Comment:
My understanding of the Ruby API is that this isn't just making the compiler happy, `rb_raise` does _not_ stop execution of the extension. Instead, it flags the VM to enter exception handling mode with this exception _when control returns to the VM_. Thus, this value is returned to Ruby, and, I think, bound to the lhs of the statement.

Reply:
This turns out to be quite a thorny issue! `rb_raise` does in fact stop execution of the C extension immediately using `longjmp` (with a corresponding `setjmp` before the extension is invoked) which is bad since it means no C++ destructors are invoked as you pop back up the stack. I think the way to fix this is to use C++ exceptions throughout most of the extension and then have very carefully crafted top level "gateway" functions that don't rely on any stack allocated object destructors and invoke `rb_raise` e.g.

```c
VALUE encode_model_with_error_handling(VALUE self, VALUE rb_model, VALUE rb_is_value_schema) {
  VALUE result = encode_model(self, rb_model, rb_is_value_schema);
  if (rb_obj_is_kind_of(result, rb_eException) == Qtrue) {
    rb_exc_raise(result);
  }
  return result;
}
```
This highlights another problem though. Any time the C extension calls back into Ruby with `rb_funcall` a similar `longjmp` will happen if that Ruby method throws an exception leading to memory leaks. The fix is use `rb_protect` so you can convert the Ruby exception to a C++ exception  that can then be handled further up the stack:

```
int state;
VALUE result = rb_protect(dangerous_func, dangerous_arg, &state);

if (state) {
  // an exception occurred
  VALUE err = rb_errinfo();
  throw RubyException(err);
}
else {
  // no exception occurred
  return result;
}
```
Gross. Fortunately the extension only makes two calls back into Ruby that could realistically throw exceptions so the problem isn't too widespread. 

All of this does make me think I should take another look at [rice](https://github.com/jasonroelofs/rice) though. I had initially ruled that library out since it I was worried the abstraction layer wouldn't really encapsulate the underlying Ruby APIs resulting in developers needing to understand both the Ruby APIs and the Rice APIs but maybe the error handling justifies this risk.

Reply:
Wow, what a mess. This makes it so easy to leak memory! Having the error handling convert C++ exceptions to Ruby exceptions as few times as possible sounds promising, as does using `rice`. I wonder how many native extensions do this wrong and just leak bits of memory here and there.